src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceParameterService.cs

```diff
@@ -0,0 +1,153 @@
+                    editor.ReplaceNode(invocationExpression, editor.Generator.InvocationExpression(invocationExpression.Expression, allArguments));
+                }
+
+                var newRoot = editor.GetChangedRoot();
+                document = document.WithSyntaxRoot(newRoot);
+                currentSolution = document.Project.Solution;
+            }
+
+            return currentSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter
+        /// </summary>
+        /// <param name="expression"> The expression containing identifiers we are tying back to the parameters </param>
+        /// <param name="methodSymbol"> The method containing the expression </param>
+        /// <returns></returns>
+        private static Dictionary<IParameterSymbol, IdentifierNameSyntax> TieExpressionToParameters(ExpressionSyntax expression, IMethodSymbol methodSymbol)
+        {
+            var nameToParameterDict = new Dictionary<IParameterSymbol, IdentifierNameSyntax>();
+            var variablesInExpression = expression.DescendantNodes().Where(node => node is IdentifierNameSyntax);
+
+            foreach (var variable in variablesInExpression)
+            {
+                foreach (var parameter in methodSymbol.Parameters)
+                {
+                    if ((string)((IdentifierNameSyntax)variable).Identifier.Value == parameter.Name)
+                    {
+                        nameToParameterDict.Add(parameter, (IdentifierNameSyntax)variable);
+                        break;
+                    }
+                }
+            }
+            return nameToParameterDict;
+        }
+
+        protected override bool ExpressionWithinParameterizedMethod(ExpressionSyntax expression)
+        {
+            var methodExpression = expression.FirstAncestorOrSelf<MethodDeclarationSyntax>(node => node is MethodDeclarationSyntax, true);
+            var variablesInExpression = expression.DescendantNodes().Where(node => node is IdentifierNameSyntax);
+            var variableCount = 0;
+            var parameterCount = 0;
+
+            foreach (var variable in variablesInExpression)
+            {
+                variableCount++;
+                foreach (var parameter in methodExpression.ParameterList.Parameters)
+                {
+                    if ((string)((IdentifierNameSyntax)variable).Identifier.Value == (string)parameter.Identifier.Value)
+                    {
+                        parameterCount++;
+                        break;
+                    }
+                }
+            }
+            return variablesInExpression.Any() && methodExpression.ParameterList.Parameters.Any() && variableCount == parameterCount;
+        }
```

Comment:
i can't tell what this method is doing.

Reply:
The idea was to determine if there are any variables within the expression, if the method has parameters, and if the variables that are in the expression tie to some parameter to filter out expressions containing local variables.

Reply:
got it.  this should be done semantically, not syntactically.

Reply:
Changed, now done semantically